drivers/char/tpm/tpm-chip.c

```diff
@@ -474,13 +474,19 @@ static void tpm_del_char_device(struct tpm_chip *chip)
 
 	/* Make the driver uncallable. */
 	down_write(&chip->ops_sem);
-	if (chip->flags & TPM_CHIP_FLAG_TPM2) {
-		if (!tpm_chip_start(chip)) {
-			tpm2_shutdown(chip, TPM2_SU_CLEAR);
-			tpm_chip_stop(chip);
+	/* Check if chip->ops is still valid since in case of power loss
```

Comment:
Is it really a power loss? May interpretation of power loss would be, that the device was unplugged.

Reply:
I just chose the wording that is used in the description of function tpm_class_shutdown()...

Reply:
Now I see where you got this from. But the comment sounds like when a power loss happened you need to check if chip-ops is still valid. But if I understand the comment in tpm-chip.c correct. The function tpm_class_shutdown should be called to prepare for a power loss. For example when the system is shutting down.
I guess a reboot would normally not result in a power loss. But to shutdown the device anyways seems to be a good idea.
My idea would be something like this:
Check if chip-ops is still valid since in case of a reboot or shutdown tpm_class_shutdown() has already sent the TPM2_Shutdown command and set it to NULL.

Reply:
Ok, I will rephrase the comment then (and omit the part about "power loss")

Reply:
But note that the terminology used in the kernel code does not seem to distinguish between "shutdown" and "reboot" (in both cases the "shutdown" handlers are called). So I think its ok to only mention "shutdown".

Reply:
Please note that beside the fixed comment there have been some more changes in the DSA part, so a (new) review of this part would be nice.